flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/parser/TransformParser.java

```diff
@@ -123,10 +125,22 @@ private static RelNode sqlToRel(
                 if (udf.getReturnTypeHint() != null) {
                     // This UDF has return type hint annotation
                     returnTypeInference =
-                            o ->
-                                    o.getTypeFactory()
-                                            .createSqlType(
-                                                    convertCalciteType(udf.getReturnTypeHint()));
+                            o -> {
+                                RelDataTypeFactory typeFactory = o.getTypeFactory();
+                                DataType returnTypeHint = udf.getReturnTypeHint();
+
+                                if (returnTypeHint instanceof ArrayType) {
+                                    DataType elementTypeHint =
+                                            ((ArrayType) returnTypeHint).getElementType();
+                                    RelDataType elementType =
+                                            typeFactory.createSqlType(
+                                                    convertCalciteType(elementTypeHint));
+                                    return typeFactory.createArrayType(elementType, -1);
+                                }
```

Comment:
#3740 also fixes UDF return type hinting issue. Can we use something similar to

```java
o -> convertCalciteType(o.getTypeFactory(), udf.getReturnTypeHint());
```

and modify `convertCalciteType` to handle `ROW`, `ARRAY`, `MAP` types correctly? Putting complex cases in a lambda expression might harm readability.